### ISSUE TITLE
Pass value from parent to subtree

### ISSUE THREAD

**Issue by dotlogix** [graphql-dotnet/graphql-dotnet#1293]
Is it somehow possible to pass an argument to a child value resolver?

As far as I understand the source code, variables are always the same instance for all resolvers, but I need to pass it for the whole subtree only. For another object, this object will be changed.

I try to achieve a nested UnitOfWork structure, so each subtree will get a new UnitOfWork.
For mutations, I want to commit all changes only after all mutations executed correctly.

To mutate arguments or variables would be a good way to achieve this, but because the engine executes everything in parallel, there is no guarantee another node wouldn't override the value in between.

So basically I need a way to provide/inject values in child nodes.

Thx in advance.

**Comment by sungam3r** [comment]
As n option you can try to achive this by `UserContext` property to store arguments in your special tree structure.

**Comment by dotlogix** [comment]
Interesting idea, but it does not provide any way to ensure access to parent context, to get proper nesting working.

I also thought about an AsyncLocal variable to apply the value to the logical async execution path, but because of the internal implementation

Pseudo Code:
ForEach Executor In Queue
-- Execute Field As Result
-- IF Result Is Object
-- Queue.Add(Result.Children)

So the hierarchy and the execution path is completely flat and not recursive so no chance to achieve sth like this.

Also, UserContext is unique and the same instance for the whole schema, so I can pass arguments in, but if I mutate them it will change for all executors and not only for children.

**Comment by sungam3r** [comment]
It is just a shared state, usage of which is up to you.

**Comment by dotlogix** [comment]
I understand, but even if I would add a custom context logic to get the correct context I still need a way to relate a child node to its parent, what is not possible right?

Her is an example:
Root
 - Field 1 (int)
 - Field 2 (string)
 - Field 3 (Object)
   - Sub Field 3.1
 - Field 4 (Object)
   - Sub Field 4.1

If you Imagine you have a context class UserContext and NestedUserContext
How could I ensure Field 1-4 get an instance of UserContext,
but SubField 3.1 gets a nested Context based on the parent
and SubField 4.1 gets a nested Context based on parent

But the Context provided for 3.1 and 4.1 is not the same instance

It gets far more complicated if you imagine there is another nested layer. Then I would expect
Root gets a Context
SubField 3.1 Gets a nested Context based on Root
SubField 3.1.1 Gets a nested Context based on 3.1

Because the Api executes everything in one layer and not in a recursive way this is just impossible

**Comment by barryhagan** [comment]
Take a look at using the `Path` property on `ResolveFieldContext`.   This is an array of names to get you from root to the current leaf.

You could use this to build keys for a concurrent dictionary to load whatever context/parent object you need based on where you are.   I would probably make a locator object that takes a `Path` and gives you back an appropriate object scoped from the parent knowing that parent's `Path` will be the current `Path` except the last item.   If that is an empty array, you are at the root.

Even though the nested mutations will run in parallel at a given depth, you should always have the parent loaded before the children fire.

**Comment by JoeEager** [comment]
Is there an example of this?  I'm looking to use the parent value to retrieve a subvalue

ParentA {
   ParentId
   ChildB {
      Value1 //Changes based on ParentId
   }
}

**Comment by sungam3r** [comment]
I think that this issue is a duplicate of #899 . May be fixed by #2295. Instead of passing values from parent to subtree you can climb up the tree from child node to its parent , grandparent and so on to get arguments or some state that you need.

**Comment by sungam3r** [comment]
@dotlogix #2295 was merged and will be published in the v4 release.
